main.tf

```
@@ -1,13 +1,117 @@
-resource "random_integer" "example" {
+locals {
+  service_account_name = coalesce(var.deployment_name, module.this.id, "tfc-agent")
+  deployment_name      = coalesce(var.deployment_name, module.this.id, "tfc-agent")
+
+  namespace = coalesce(var.kubernetes_namespace, "default")
+}
+
+resource "kubernetes_namespace" "namespace" {
+  count = var.namespace_creation_enabled ? 1 : 0
+  metadata {
+    name = local.namespace
+  }
+}
+
+resource "kubernetes_service_account" "service_account" {
   count = module.this.enabled ? 1 : 0
 
-  min = 1
-  max = 50000
-  keepers = {
-    example = var.example
+  metadata {
+    name        = local.service_account_name
+    namespace   = local.namespace
+    annotations = var.service_account_annotations
   }
 }
 
-locals {
-  example = format("%v %v", var.example, join("", random_integer.example[*].result))
+resource "kubernetes_secret" "secret" {
+  metadata {
+    name      = local.deployment_name
+    namespace = local.namespace
+  }
+
+  data = {
+    token = var.tfc_agent_token
+  }
+}
+
+resource "kubernetes_deployment" "tfc_cloud_agent" {
+  count = module.this.enabled ? 1 : 0
+
+  metadata {
+    name      = local.deployment_name
+    namespace = local.namespace
+    labels    = module.this.tags
+  }
+  spec {
+    selector {
+      match_labels = module.this.tags
+    }
+    replicas = var.replicas
+
+    template {
+      metadata {
+        namespace   = local.namespace
+        labels      = module.this.tags
+        annotations = var.deployment_annotations
+      }
+      spec {
+        service_account_name            = local.service_account_name
+        automount_service_account_token = true
+        container {
+          image = var.agent_image
+          name  = "tfc-agent"
+          args  = var.agent_cli_args
+          env {
+            name = "TFC_AGENT_TOKEN"
+            value_from {
+              secret_key_ref {
+                key  = "token"
+                name = local.deployment_name
+              }
+            }
+          }
+          env {
+            name  = "TFC_AGENT_NAME"
+            value = coalesce(module.this.id, "tfc-agent")
+          }
+          env {
+            name  = "TFC_AGENT_LOG_LEVEL"
+            value = var.tfc_agent_log_level
+          }
+          env {
+            name  = "TFC_AGENT_DATA_DIR"
+            value = var.tfc_agent_data_dir
+          }
+          env {
+            name  = "TFC_AGENT_SINGLE"
+            value = var.tfc_agent_single
+          }
+          env {
+            name  = "TFC_AGENT_DISABLE_UPDATE"
+            value = var.tfc_agent_disable_update
+          }
+          env {
+            name  = "TFC_ADDRESS"
+            value = var.tfc_address
+          }
+          dynamic "env" {
+            for_each = var.tfc_extra_envs
```

I think the `tfc_....` prefix makes sense for the envs like `TFC_ADDRESS` since those are actually what the canonical environment variables are named. Since everything in this module is about TFC, I think adding the additional `tfc_` prefix is superfluous and we don't prefix many other vars with it. 
```suggestion
            for_each = var.envs
```

(or `extra_envs`, both okay)

---

or maybe `agent_envs` since you have `agent_image`, `agent_cli_args`...